0010-R-components-healthcheck.md

```diff
@@ -0,0 +1,477 @@
+        },
+        {
+            "componentName": "mockABC",
+            "type": "pubsub",
+            "status": "UNDEFINED",
+            "errorCode": "ERR_PING_NOT_IMPLEMENTED"
+        }
+}
+```
+
+**Case 4:** When ANY component in system DOES NOT implement Ping: Here we report ALL those components for which Ping is not implemented.
+
+HTTP Response code: 405 Method Not Allowed
+GRPC Response code (if implemented): 12 UNIMPLEMENTED
+
+```
+{
+    "componentsHealth": 
+        {
+            "componentName": "txnstore",
+            "type": "state",
+            "status": "UNDEFINED",
+            "errorCode": "ERR_PING_NOT_IMPLEMENTED"
+        },
+        {
+            "componentName": "smsbinding",
+            "type": "bindings",
+            "status": "UNDEFINED",
+            "errorCode": "ERR_PING_NOT_IMPLEMENTED"
+        },
+        {
+            "componentName": "mockABC",
+            "type": "pubsub",
+            "status": "UNDEFINED",
+            "errorCode": "ERR_PING_NOT_IMPLEMENTED"
+        },
+        {
+            "componentName": "mockXYX",
+            "type": "bindings",
+            "status": "UNDEFINED",
+            "errorCode": "ERR_PING_NOT_IMPLEMENTED"
+        }
+}
+```
+**Case 5:** When SOME components in system implement Ping and whatever Compoenents implement Ping are Healthy:
+
+HTTP Response code: 200 OK
+GRPC Response code (if implemented): 0 OK
+
+```
+{
+    "componentsHealth":
+        {
+            "componentName": "mockABC",
+            "type": "pubsub",
+            "status": "UNDEFINED",
+            "errorCode": "ERR_PING_NOT_IMPLEMENTED"
+        },
+        {
+            "componentName": "mockXYX",
+            "type": "bindings",
+            "status": "UNDEFINED",
+            "errorCode": "ERR_PING_NOT_IMPLEMENTED"
+        }
+}
+```
+**Pros of this Approach:**
+1. Consistent view of result and only Required components are listed.
+
+
+**Cons of this Approach:**
+1. May look repetitive for Failed and Ping Not implemented components, and user may need to search specifically for failed components, when components list for "Ping Not Implemented" would be huge.
+
+### Approach XYZ:
+
+There can be other approaches, with different kind of endpoints. Or a dedicated endpoint, like: `http://localhost:3500/v1.0-alpha1/healthz/component`
+
+
+### RECOMMENDED APPROACH
+Approach 3, as it seems to be a Hybrid of Approach 1 and Approach 2. It only reports required components and keeps a consistent view.
+And, to implement only http endpoint.
+
+### All Approaches with Question:
+Should we support query for a partular component Health check?
+I don't think that is required specifically, as of now.
+It is more about overall health of the system.
+
+
+## Open Questions
+1. Is it only for kubernetes users? Is it only needed for http endpoint? Or we should cover gRPC endpoint as well?
```

Comment:
Comment from Alessandro (@ItalyPaleAle ): No, although K8s is arguably the main user of this.

Ref: https://hackmd.io/MaNpUYRyQqe-0eqCsSUiIg

Reply:
Comment from Alessandro (@ItalyPaleAle ): Which brings up a question: should we add an annotation to indicate which components to include in the K8s healthchecks?

Ref: https://hackmd.io/MaNpUYRyQqe-0eqCsSUiIg